src/hotspot/share/opto/divconstants.cpp

```diff
@@ -0,0 +1,210 @@
+//
+// (a) For positive values c, m such that:
+//
+// m <= c * d < m + m / v
+//
+// We have:
+//
+// floor(x / d) = floor(x * c / m) for every integer x in [0, N]
+//
+// (b) For positive values c, m such that:
+//
+// m < c * d <= m + m / v
+//
+// We have:
+//
+// ceil(x / d) = floor(x * c / m) + 1 for every integer x in [-N, 0)
+//
+// (2) Proof:
+//
+// (a) The conclusion is trivial for x = 0
+//
+// For 0 < x <= v
+//
+// Since 1 / d <= c / m < (1 / d) * ((v + 1) / v)
+//
+// We have x / d <= x * c / m < (x * ((v + 1) / v)) / d
+//
+// As a result, since x * ((v + 1) / v) <= x * ((x + 1) / x) = x + 1
+//
+// x / d <= x * c / m < (x + 1) / d, which implies floor(x / d) = floor(x * c / m) since
+// there can be no integer in (x / d, (x + 1) / d)
+//
+// For v + 1 <= x <= v + d - 1, since v >= d - 1, we have x <= 2v
+// As a result, x * ((v + 1) / v) <= x * ((x + 2) / x)
+//
+// floor(x / d) = (v + 1) / d
+// x * c / m < x * ((v + 1) / v) / d) <= x * ((x + 2) / x) / d = (x + 2) / d
+//           <= (v + d - 1 + 2) / d = (v + 1) / d + 1
+// Which means (v + 1) / d <= floor(x * c / m) < (v + 1) / d + 1 with (v + 1) / d being an integer
+// This implies floor(x / d) = floor(x * c / m) for v + 1 <= x <= v + d - 1
+//
+// Combining all the cases gives us the conclusion.
+//
+// (b) Since ceil(a / b) = floor((a - 1) / b) + 1, we need to prove:
+//
+// floor((x - 1) / d) = floor(x * c / m)
+//
+// For 0 > x >= -v
+//
+// Since 1 / d < c / m <= (1 / d) * ((v + 1) / v)
+//
+// We have x / d > x * c / m >= (x / d) * ((v + 1) / v)
+//
+// since x * ((v + 1) / v) >= x * ((x - 1) / x) = x - 1
+// x / d > x * c / m >= (x - 1) / d, which implies floor((x - 1) / d) = floor(x * c / m) since
+// there can be no integer in ((x - 1) / d, x / d)
+//
+// For -v - d + 1 <= x <= -v - 1, since v >= d - 1, we have x >= -2v
+// As a result, x * ((v + 1) / v) >= x * ((x - 2) / x) = x - 2
+//
+// x / d <= (-v - 1) / d
+// floor((x - 1) / d) = (-v - 1) / d - 1
+// x * c / m >= x * ((v + 1) / v) / d) >= (x - 2) / d
+//           >= (-v - d + 1 - 2) / d = (-v - 1) / d - 1
+// which means (-v - 1) / d >= x / d > x * c / m >= (-v - 1) / d - 1
+// This implies floor((x - 1) / d) = floor(x * c / m) for -v - 1 >= x >= -v - d + 1
+//
+// Combining all the cases gives us the conclusion.
+//
+// (3) Discussion:
+//
+// Let x be v, v - d + 1, -v, -v + d - 1, it can be seen that these bounds are indeed optimal
+//
+// (4) Implementation:
+//
+// For computation efficiency, we only consider the values m = 2**s.
+// This function does not handle the cases d being a power of 2, which means
+// that c * d is never equal to m.
+//
+// We find the value of c, m such that it satisfies the bounds for both the nonnegative
+// and negative ranges of x. This can be done by finding v_neg and v_pos and the bounds of
+// c * d - m is the intersection of (0, m / v_neg] and (0, m / v_pos). Which is (0, m / v_pos)
+// if v_pos >= v_neg and (0, m / v_neg] otherwise.
+//
```

Comment:
The analysis seem correct.

Reply:
The convoluted test to break out of the while loop can be simplified if we are willing to consider two cases.

Equality in `c / m <= (1 / d) * ((v + 1) / v)` is equivalent to `c / m = b / v`, where `b = (v + 1) / d` is an integer. Note that the fraction `b / v` is irreducible: a common divisor of `b` an `v` has to divide any linear combination as well, in particular it has to divide `d * b - 1 * v = 1`. Thus, `c >= b` and `m >= v`. Since `m = 2^s`, `v` must be a power of 2.

This means that, when `m` is a power of 2, the only way to have equality is `v_neg > v_pos` _and_ `v_neg` is a power of 2 (say `v_neg = 2^e`), which should be a rare case. This can be detected cheaply early, before the loop. In such a  case, the smallest `c` is `b` and the smallest `s` is `e`, and a result meeting `s >= min_s` doesn't need any iterative algorithm.

Otherwise equality does not hold. Now, `c / m < (1 / d) * ((v + 1) / v)` is equivalent to `m > v * rc`. In turn, `m > v * rc` <=> `m / v > rc` <=> `ceil(m / v) > rc`.

Thus, rather than maintaining invariants for `qv = floor(m / v)` and `rv = m - qv * v` as currently defined, we can redefine them as `qv = ceil(m / v)` and `rv = qv * v - m` (`0 <= rv < v`) and maintain _these_ invariants instead.
```
  T qv = 1;
  T rv = v - 1;
    ...
    if (rv >= v - rv) { // 2 * rv >= v
      qv_ovf = qv > min_signed;
      qv = qv * 2 - 1;
      rv = rv * 2 - v;
    } else {            // 2 * rv < v
      qv_ovf = qv >= min_signed;
      qv = qv * 2;
      rv = rv * 2;
    }
```
The test to exit the loop then reduces to `qv > rc || qv_ovf`.

Reply:
That is an excellent analysis. To add to the analysis, we do not really need the minimal value of `c`, since 2 values of `c` that both satisfy the inequations must give the same upper bits for all input values. As a result, for the purpose of the algorithm, they are equivalent.

My concern is that it will complicate the analysis, which is complicated enough, for a minor improvement in the exit conditions.

Reply:
I have no idea about the timing difference with the current exit condition or with the simplified one. It might indeed be negligible.
Anyway, there's a choice now ;-)

Reply:
@rgiulietti Thanks a lot for your patience in reviewing this patch. Do you have any more concerns?

Reply:
I've the impression that we can replace `m < c * d <= m + m / v` with the stricter `m < c * d < m + m / v` by using `N_neg - 1` instead of `N_neg`, but I need some time to have a solid proof.

That would simplify the code of the algorithm.

Reply:
But IMO the current algorithm is correct.

Reply:
Nope, my impression [here](https://github.com/openjdk/jdk/pull/9947#discussion_r1465362000) turns out to be wrong.